doc/compiler/il/SymbolsSymrefsAliasing.md

```diff
@@ -0,0 +1,171 @@
+# Symbols, Symbol References, and Aliasing in Testarossa
+
+## Introduction
+
+This document explains and documents the concepts and implementation of symbols,
+symbol references, and aliasing in Testarossa.
+
+## Symbols and Symbol References
+
+The original, conceptual distinction between symbols and symbol references (symrefs)
+was actually quite straightforward. However, over time, implementation details have
+led to the definition of the two to become very blurry and somewhat unspecified.
+This document therefore offers the original definitions, as well as some implementation
+details and subtleties.
+
+### Symbols
+
+A symbol represents a memory location. It holds the minimum information required to
+directly translate a method's source code or IL representation, without consideration
+for optimization. It stores basic information (e.g. data type) needed to determine
+what memory access operations can/must be done.
+
+Different kinds of symbols exist to represent the different kinds of memory access
+requirements. Each kind of symbol stores information that is specific to its kind (type).
+The different symbol kinds are:
+
+* static
+   * represents a static memory address
+   * contains the address of the memory location
+* auto
+   * represents a variable mapped on the stack
+      * may come from local method variables or compiler allocated stack objects
+   * contains the stack offset of the variable
+* method
+   * represents a method (static, virtual)
+* resolved method
+* parameter
+* register mapped
+* label
+* shadow
+   * represents indirection off a base address
+      * e.g. accessing a class field using the class base address plus an offset
+   * contains offset from some base address
+
+#### int shadow
+
+An int shadow represents the common case of an integer type. In the case of Java,
+for example, an int shadow is created for a field or array access.
+
+#### generic int shadow
+
+A generic int shadow is created by the optimizer in (relatively rare) to refer to a
+field that wasn't referred to in the bytecodes. If the optimizer wants to synthesize
+a load or store, but doesn't have a symref for it in the original program, it can
```

Comment:
You mean the optimizer doesn't have a symref or symbol? If it is symref, you should spell out symref as symbol reference.